### ISSUE TITLE
Strange hanging issue

### ISSUE THREAD

**Issue by k-bx** [actix/actix-web#2443]
Hi. I'm using Actix for a somewhat high-load server that's running on four Ubuntu 20.04 machines behind Amazon Application Balancer. They sometimes get in a strange "lock" state where they suddenly hang on some requests. This might be not specific to Actix in any form, but there are now details which might give Actix maintainers some helpful hints which I would very much appreciate.

So, here's what's given: there's actix web server with a bunch of endpoints, a static endpoint set up as this:

```rust
            .service(fs::Files::new("/static", "./static"))
```

and another endpoint setup like this:

```rust
            .route(
                "/markup-inspector",
                web::get().to(crate::markup_inspector::markup_inspector),
            )
// ...
pub async fn markup_inspector(
    auth: actix_web_httpauth::extractors::basic::BasicAuth,
) -> Result<web::HttpResponse, AppError> {
    println!("> debug: markup_inspector: before auth");
    crate::auth::auth_basic(auth)?;
    println!("> debug: markup_inspector: before serving index.html");
    Ok(NamedFile::open("static/markup-inspector/index.html")?)
}
```

What happens is, over time these two faulty web servers (or one, or three, randomly) would get in a state where this endpoint would print the messages, but would then become less responsive (35sec response), and even hang.

```
ubuntu@ip-10-100-11-12:~$ curl -i localhost:8080/markup-inspector | head -n20
// hangs
```

Restarting the server doesn't help, it quickly gets into a similar situation. Rebooting does help.

But! What is interesting is that I've reimplemented this endpoint to read the file contents before responding:

```rust
pub async fn markup_inspector(
    auth: actix_web_httpauth::extractors::basic::BasicAuth,
) -> Result<web::HttpResponse, AppError> {
    println!("> debug: markup_inspector: before auth");
    crate::auth::auth_basic(auth)?;
    println!("> debug: markup_inspector: before serving index.html");
    let bytes =
        async_std::fs::read("static/markup-inspector/index.html").await?;
    Ok(web::HttpResponse::Ok()
        .insert_header((
            "content-disposition",
            "inline; filename=\"index.html\"",
        ))
        .body(bytes))
}
```

This did solve the problem for the endpoint, however the issue is still happening for the `/static` route. So, when you redeploy without reboot, you'll be able to query the `localhost:8080/markup-inspector` endpoint, while anything from `/static` will hang.

I'm very puzzled by the issue, but thought that this distinction between using and not using the `NamedFile` might enable actix devs to help me out by giving some ideas on what might be happening and what should I check.

The limits for the server in terms of open files are quite high, but maybe there's something else I'm missing.

```
ubuntu@ip-10-100-10-166:~$ cat /proc/1027552/limits 
Limit                     Soft Limit           Hard Limit           Units     
Max cpu time              unlimited            unlimited            seconds   
Max file size             unlimited            unlimited            bytes     
Max data size             unlimited            unlimited            bytes     
Max stack size            8388608              unlimited            bytes     
Max core file size        0                    unlimited            bytes     
Max resident set          unlimited            unlimited            bytes     
Max processes             59948                59948                processes 
Max open files            65536                65536                files     
Max locked memory         65536                65536                bytes     
Max address space         unlimited            unlimited            bytes     
Max file locks            unlimited            unlimited            locks     
Max pending signals       59948                59948                signals   
Max msgqueue size         819200               819200               bytes     
Max nice priority         0                    0                    
Max realtime priority     0                    0                    
Max realtime timeout      unlimited            unlimited            us        
```

versions:

```
rustc 1.58.0-nightly (495322d77 2021-11-08)
actix-web v4.0.0-beta.9
actix-files v0.6.0-beta.7
```

Thank you!

**Comment by robjtede** [comment]
This is a strange issue indeed.
To help with debugging, can you also list the version of `actix-http` and `actix-server` that are being used?

**Comment by k-bx** [comment]
Sure, it's `actix-http v3.0.0-beta.11` and `actix-server v2.0.0-beta.6`. Attaching the whole `cargo tree` output just in case I've missed something

[deps.txt](https://github.com/actix/actix-web/files/7538246/deps.txt)

**Comment by fakeshadow** [comment]
try use tokio::fs instead of async_std and see what happens. In fact you should not use async_std in actix-web app when you can.

to be clear I'm not saying tokio::fs would solve the problem. It's that tokio::fs share the same runtime detail with actix-files. Namely the same blocking pool for file operation. Using it would make your debug easier as we can roll out one possibility(or nail the cause who knows)

**Comment by k-bx** [comment]
@fakeshadow that is an interesting point, didn't know that, thank you!

I've removed the async_std dependency completely from our codebase, will observe the effect.

**Comment by k-bx** [comment]
Ok, so once I've switched off the async_std, it does hang even in the case of `tokio::fs::read("static/markup-inspector/index.html").await?;`. So, probably this is specific to tokio runtime, but I'm adding a print right before `Ok(web::HttpResponse::Ok()...` just to confirm it's hanging on reading the FS.

**Comment by robjtede** [comment]
Can you try increasing [the blocking worker count](https://docs.rs/actix-web/4.0.0-beta.11/actix_web/struct.HttpServer.html#method.worker_max_blocking_threads). It's 512 / workers by default so try doubling it and see.

**Comment by k-bx** [comment]
Thanks, deploying this change. I should also note that the hang is now at least consistent between the static endpoint and my manual one. But, I haven't checked whether it hangs right before returning the result or when it's doing `tokio::fs::read`, I've added another print just in case it hangs again.

**Comment by 0xpr03** [comment]
Just a an idea: you can also use [tracing](https://docs.rs/tracing/0.1.29/tracing/index.html) to get more data out of tokio, though it'll probably spam a lot without fine tuning. And if you want you can look into the tokio [console](https://github.com/tokio-rs/console) ([beta announcement](https://tokio.rs/blog/2021-09-console-dev-diary-1)) which gives you a htop like view of the processes. Problem is that you probably don't want too much load on the inspected process, or things will be all over the place very fast.

**Comment by k-bx** [comment]
Hey everyone. Apologies for the delay. Just wanted to mention that the blocking worker count helped, systems are now much more stable and definitely not seeing the previous issue. There are some irregularities still, and I've started using tokio-console, but probably need to find some ways to annotate processes better. Will keep gradually improving my understanding around the tooling and maybe even contribute.

Closing this issue now. Huge thanks for help, once again!
